[REVIEW] contiguous_split function.

nvdbaranec opened the pull request:
Implementation of contiguous_split() functionality.   Performs a deep-copying split operation on a table_view where the output is arranged in a memory-contiguous way per output table-view.  That is, each output table_view and all associated internal column_views share one contiguous chunk of allocated memory.  Thus, these various objects are not owned by top level tables or columns and have a special lifetime controlled by the returned contiguous_split_result objects.  


The basic algorithm is structured as:

```
for(each input column in a table){
   split_table_subcolumns[i] = slice(in.column(i), splits);
}
for(each output table i){
    contiguous_buf_size = 0;
    for(each subcolumn j){
       contiguous_buf_size += compute(split_table_columns[j][i]);
    }
    buf = alloc(contiguous_buf_size);
    for(each subcolumn j){
        // buf gets incremented each call
        copy_column(buf, split_table_columns[j][i]);
    }
}
```
O(numsplits) allocations and O(numsplits * numcolumns) copies. 

The underlying column views returned from this function have UNKNOWN_NULL_COUNT as a byproduct of the way cudf::split() and cudf::slice() work.   There is an open issue for this  (https://github.com/rapidsai/cudf/issues/3600).  Once fixed, there's a very simple change that can be made to contiguous_split() to propagate the actual null counts.

harrism commented:
> I would imagine this lives in copy.cu at the end of the day.

Food for thought: The more that goes in a single .cu file the longer that file takes to compile and the less parallel efficiency we get in our builds.

harrism commented:
Highly recommend building benchmarks using gbenchmark rather than ad-hoc so that we can put them into our benchmark suite.

jrhemstad commented:
I highly suggest that Google Benchmarks be created for performance testing before anything else is done.

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,550 @@
+   std::vector<subcolumns> split_table;
+   split_table.reserve(input.num_columns());
+   for(int idx=0; idx<input.num_columns(); idx++){
+      split_table.push_back(cudf::experimental::slice(input.column(idx), splits));
```
This looks like it could be replace with the zero-copy `slice` API that already exists: https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/include/cudf/copying.hpp#L325

jrhemstad replied:
Furthermore, the zero-copy `slice` API should be extended to provide a `table_view` API to do a zero-copy slice of a table.

nvdbaranec replied:
Just to clarify, the above code is using the zero-copy slice.   You're saying :  make a zero-copy slice function that works on table views?    

Did you want it to be split or slice?  At the moment, split is just a wrapper around slice that takes a different form for the input.

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,550 @@
+                           // PERFORMANCE TEST 2, 3, 4
+                           /*, nullptr*/);
+      }
+      mcv.set_null_count(cudf::UNKNOWN_NULL_COUNT);
```
This isn't doing anything. This modifies a function local object which then is destroyed.

nvdbaranec replied:
It is.  That mcv gets converted to a column_view and put on the output list.

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,550 @@
+
+template<typename T>
+struct column_buf_size_functor_impl {
+   void operator()(column_view const& c, size_t& running_data_size, size_t& running_validity_size)
```
Don't use pass by reference for outputs. Outputs should be _returned_.

nvdbaranec replied:
But in this case, it's only filling in part of a vector.  Doesn't it make more sense to have a single vector that gets filled in with multiple function calls rather than allocating mutiple vectors as return values and stitching them together?

jrhemstad replied:
See my comment about using `std::accumulate` in `allocate_and_copy`. This function would be made to return a `std::pair<std::size_t, std::size_t>`.

nvdbaranec commented:
Did a whole bunch of refactoring and added table_view versions of slice() and split() at Jake's suggestion.   I put together some benchmarks using googlebenchmarks.  

Summary below.   Ran with various combinations of  
- 6 gigs / 1 gig total data size
- 512 columns / 10 columns
- validity / no validity data

-----------------------------------------------------------------------------
Benchmark            Time      CPU     Iterations UserCounters...
-----------------------------------------------------------------------------
6Gb512ColsNoValidity 448 ms    448 ms  20         bytes_per_second=13.3939G/s
6Gb512ColsValidity   451 ms    450 ms  20         bytes_per_second=13.317G/s
6Gb10ColsNoValidity  61.6 ms   61.6 ms 20         bytes_per_second=97.4225G/s
6Gb10ColsValidity    63.9 ms   64.0 ms 20         bytes_per_second=93.8442G/s
1Gb512ColsNoValidity 441 ms    441 ms  20         bytes_per_second=2.26546G/s
1Gb512ColsValidity   443 ms    443 ms  20         bytes_per_second=2.25595G/s
1Gb10ColsNoValidity  17.2 ms   17.3 ms 20         bytes_per_second=58.0719G/s
1Gb10ColsValidity    18.0 ms   18.0 ms 20         bytes_per_second=55.6857G/s

As expected, this is highly bound by total # of output columns (input columns * num splits). Primary optimization is probably a unifier per-split kernel that copies multiple columns.

jrhemstad commented:
```diff
diff --git a/cpp/include/cudf/detail/copy.hpp b/cpp/include/cudf/detail/copy.hpp
index b1a6e0e..570b56f 100644
--- a/cpp/include/cudf/detail/copy.hpp
+++ b/cpp/include/cudf/detail/copy.hpp
@@ -79,6 +79,8 @@ ColumnView slice(ColumnView const& input,
  * @note It is the caller's responsibility to ensure that the returned view
  * does not outlive the viewed device memory contained in the `all_data` field of the
  * returned contiguous_split_result.   
+ * 
+ * test
  *
  * Example:
  * input:   [{10, 12, 14, 16, 18, 20, 22, 24, 26, 28},

```

jrhemstad commented:
Test. Ignore. 
```diff
diff --git a/cpp/include/cudf/detail/copy.hpp b/cpp/include/cudf/detail/copy.hpp
index b1a6e0e..570b56f 100644
--- a/cpp/include/cudf/detail/copy.hpp
+++ b/cpp/include/cudf/detail/copy.hpp
@@ -79,6 +79,8 @@ ColumnView slice(ColumnView const& input,
  * @note It is the caller's responsibility to ensure that the returned view
  * does not outlive the viewed device memory contained in the `all_data` field of the
  * returned contiguous_split_result.   
+ * 
+ * test
  *
  * Example:
  * input:   [{10, 12, 14, 16, 18, 20, 22, 24, 26, 28},

```

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,219 @@
+   template<typename T>   
+   void operator()(column_view const& in, char*& dst, std::vector<column_view>& out_cols)
+   {
+      column_copy_functor_impl<T> fn{};      
```
What's the purpose of `column_copy_functor_impl`? Seems like an unnecessary level of indirection. Instead, the body of `column_copy_functor_impl` could be moved here and the `column_copy_functor::operator()` could be specialized accordingly.

nvdbaranec replied:
My thinking here is that the implementation for string_view is going to be different.  It's going to have to allocate multiple buffers and distribute them to child columns_views.  This is the pattern I've seen when dealing with a situation like this (eg, Dave Wendt's string_view addition to copy_if_else).

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,219 @@
+   size_t validity_size = 0;      
+
+   // compute sizes
+   std::for_each(t.begin(), t.end(), [&data_size, &validity_size](cudf::column_view const& c){
```
How about a `std::accumulate` with a `std::pair<std::size_t,std::size_t>` accumulator? `column_buf_size_functor::operator()` would be made to return a `std::pair<std::size_t, std::size_t>` as well.

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,219 @@
+   template<typename T>
+   void operator()(column_view const& c, size_t& running_data_size, size_t& running_validity_size)
+   {
+      column_buf_size_functor_impl<T> sizer{};
```
Same comment here. `column_buf_size_functor::operator()` can be specialized, which obviates the need for the extra indirection of `column_buf_size_functor_impl`.

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,207 @@
+template <size_type block_size, typename T, bool has_validity>
+__launch_bounds__(block_size)
+__global__
+void copy_in_place_kernel( column_device_view const in,
```
So I spent some time thinking about this kernel, and I think we can avoid it. 

The data portion of the columns is easy, that's just memcpys. The tricky part is the bitmasks because you can't memcpy at the granularity of individual bits. Luckily, there is already a kernel that can handle offset bitmask copies: https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/src/bitmask/null_mask.cu#L188

I think this kernel combined with simple memcpys could eliminate the need for this kernel (and therefore eliminate the need for creating `column_device_view`s and the issue with `null_count()).

nvdbaranec replied:
Ok, so I gave this a try, and while it works, it is considerably slower in the case where we actually have validity. 

```
     cudaMemcpy(data, in.data<T>(), data_size, cudaMemcpyDeviceToDevice);
      if(in.nullable()){           
         _copy_bitmask(validity, in.null_mask(), in.offset(), in.offset() + in.size(), (cudaStream_t)0);
      }
```
I made a local copy of _copy_bitmask that takes destination bitmask as an input (instead of allocating itself).

Here's the key timings (timings done using googlebenchmarks):

(Existing PR method)
6Gb, 512 cols, 256 splits, no validity :  364 ms  (avg over 20 iterations)
6Gb, 512 cols, 256 splits, validity      :  370 ms  (avg over 20 iterations)

(New method)
6Gb, 512 cols, 256 splits, no validity :  438 ms  (avg over 20 iterations)
6Gb, 512 cols, 256 splits, validity      :  767 ms  (avg over 20 iterations)


It makes sense if you think about it.  With the PR as is, we're doing 1 kernel call per copy, regardless of the presence of validity.   With this suggested method, we're doing 2 (I'm assuming cudaMemcpy() ends up effectively a kernel).  Benchmarks have always indicated that we're bound by total # of kernel calls, so doubling them exaggerates this.

nvdbaranec commented:
Apologies for leaving it with the title "WIP".  I had posted it that way so we could sync up on the core idea after you guys finished shipping 0.11.

nvdbaranec commented:
rerun tests

harrism commented:
@nvdbaranec can you update the top-level comment to serve as an up-to-date description of the PR? Still says "demo".

harrism commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,197 @@
+   size_t validity_size = sizes.second;   
+
+   // allocate 
+   auto device_buf = std::make_unique<rmm::device_buffer>(rmm::device_buffer{data_size + validity_size, stream, mr});   
```
std::make_unique invokes a constructor with the given arguments. So if it is given a whole object as here, it invokes the copy constructor. Instead:
```suggestion
   auto device_buf = std::make_unique<rmm::device_buffer>(data_size + validity_size, stream, mr);   
```

jrhemstad replied:
Just for reference, this would have invoke the move ctor, so it wouldn't have been the end of the world, but the suggested change is better and more direct.

harrism commented on cpp/tests/copying/slice_tests.cu:
```diff
@@ -173,3 +137,127 @@ TEST_F(SliceCornerCases, NegativeOffset) {
+
+TYPED_TEST_CASE(SliceTableTest, cudf::test::NumericTypes);
+
+TYPED_TEST(SliceTableTest, NumericColumnsWithInValids) {
```
```suggestion
TYPED_TEST(SliceTableTest, NumericColumnsWithNulls) {
```

harrism commented on cpp/tests/copying/slice_tests.cu:
```diff
@@ -173,3 +137,127 @@ TEST_F(SliceCornerCases, NegativeOffset) {
+
+struct SliceStringTableTest : public SliceTableTest <std::string>{};
+
+TEST_F(SliceStringTableTest, StringWithInvalids) {    
```
```suggestion
TEST_F(SliceStringTableTest, StringWithNulls) {    
```

harrism commented on cpp/tests/copying/split_tests.cu:
```diff
@@ -208,8 +191,256 @@ TEST_F(SplitCornerCases, NegativeValue) {
+
+struct SplitStringTableTest : public SplitTest <std::string>{};
+
+TEST_F(SplitStringTableTest, StringWithInvalids) {
```
```suggestion
TEST_F(SplitStringTableTest, StringWithNulls) {
```

harrism commented on cpp/tests/copying/split_tests.cu:
```diff
@@ -208,8 +191,256 @@ TEST_F(SplitCornerCases, NegativeValue) {
+    }); 
+}
+
+struct SplitStringTableTest : public SplitTest <std::string>{};
```
Shouldn't this be up with the other split() tests?

nvdbaranec commented:
Done.

________________________________
From: Mark Harris <notifications@github.com>
Sent: Monday, December 16, 2019 6:56 PM
To: rapidsai/cudf <cudf@noreply.github.com>
Cc: Dave Baranec <dbaranec@nvidia.com>; Mention <mention@noreply.github.com>
Subject: Re: [rapidsai/cudf] [REVIEW] contiguous_split function. (#3557)


@nvdbaranec<https://github.com/nvdbaranec> can you update the top-level comment to serve as an up-to-date description of the PR? Still says "demo".

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/rapidsai/cudf/pull/3557?email_source=notifications&email_token=ANQRY6XDVT2AMTVPWX3BJ7TQZAPUNA5CNFSM4JYBEA5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHAWCXY#issuecomment-566321503>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANQRY6UP3AIHNTGCKUA6NSDQZAPUNANCNFSM4JYBEA5A>.

-----------------------------------------------------------------------------------
This email message is for the sole use of the intended recipient(s) and may contain
confidential information.  Any unauthorized review, use, disclosure or distribution
is prohibited.  If you are not the intended recipient, please contact the sender by
reply email and destroy all copies of the original message.
-----------------------------------------------------------------------------------

jrhemstad commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,197 @@
+   std::pair<size_t, size_t> operator()(std::pair<size_t, size_t> sizes, column_view const& c)
+   {
+      size_t data_size = sizes.first + (cudf::util::div_rounding_up_safe(c.size() * sizeof(T), split_align) * split_align);
+      size_t validity_size = sizes.second + (c.nullable() ? cudf::bitmask_allocation_size_bytes(c.size(), split_align) : 0);
```
Yeah, this should preserve the 64B allocation quantum.

nvdbaranec commented on cpp/include/cudf/copying.hpp:
```diff
@@ -355,6 +386,84 @@ std::vector<column_view> slice(column_view const& input,
+   cudf::table_view                    table;
+   std::unique_ptr<rmm::device_buffer> all_data;
+
+   operator table_view() const { return table; };
```
It allows templating of the split() test code for re-use with contiguous_split().  Their outputs are similar in type (vector of table views in the split() case(), vector of contiguous_split_results (which are just wrapped table views) for the contiguous_split() case) so the case allows expect_tables_equal() to work on both output types.

I can refactor a bit to remove it.

nvdbaranec commented on cpp/include/cudf/detail/copy.hpp:
```diff
@@ -63,6 +63,46 @@ ColumnView slice(ColumnView const& input,
                      input.offset() + begin, children);
 }
 
+/**
```
Ah.  I definitely like this.

nvdbaranec commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -0,0 +1,197 @@
+   template <typename T, std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+   std::pair<size_t, size_t> operator()(std::pair<size_t, size_t> sizes, column_view const& c)
+   {
+      size_t data_size = sizes.first + (cudf::util::div_rounding_up_safe(c.size() * sizeof(T), split_align) * split_align);
```
Hah. Yes, what I have there is kind of dumb.

harrism commented on cpp/benchmarks/copying/contiguous_split_benchmark.cu:
```diff
@@ -24,42 +24,45 @@
-   std::vector<cudf::test::fixed_width_column_wrapper<int>> src_cols = {};
+   std::vector<cudf::test::fixed_width_column_wrapper<int>> src_cols(num_cols);
    for(int idx=0; idx<num_cols; idx++){
       auto rand_elements = cudf::test::make_counting_transform_iterator(0, [](int i){return rand();});
```
This can be outside the loop.

harrism commented on cpp/benchmarks/copying/contiguous_split_benchmark.cu:
```diff
@@ -24,42 +24,45 @@
    auto valids = cudf::test::make_counting_transform_iterator(0, [](auto i) { return true; });
-   std::vector<cudf::test::fixed_width_column_wrapper<int>> src_cols = {};
+   std::vector<cudf::test::fixed_width_column_wrapper<int>> src_cols(num_cols);
    for(int idx=0; idx<num_cols; idx++){
```
`std::generate()` would be nicer...

harrism commented on cpp/benchmarks/copying/contiguous_split_benchmark.cu:
```diff
@@ -24,42 +24,45 @@
    srand(31337);
    auto valids = cudf::test::make_counting_transform_iterator(0, [](auto i) { return true; });
-   std::vector<cudf::test::fixed_width_column_wrapper<int>> src_cols = {};
+   std::vector<cudf::test::fixed_width_column_wrapper<int>> src_cols(num_cols);
```
This could just be a vector of `cud:column` -- just use the wrapper as a convenience for creating the columns.

harrism commented on cpp/benchmarks/copying/contiguous_split_benchmark.cu:
```diff
@@ -24,42 +24,45 @@
-   }
+   }      
+   std::vector<std::unique_ptr<cudf::column>> columns(num_cols);
+   std::transform(src_cols.begin(), src_cols.end(), columns.begin(), [](cudf::test::fixed_width_column_wrapper<int> &in){   
```
can use auto for the lambda parameter type to simplify / make this more generic.  And now that `src_cols` is a vector of columns rather than wrappers, you can eliminate the src_cols and just use a for_each whose lambda only sets the null count on its inputs -- one liner.

harrism commented on cpp/src/copying/contiguous_split.cu:
```diff
@@ -80,16 +94,22 @@ struct column_buf_size_functor {
-      size_t data_size = sizes.first + (cudf::util::div_rounding_up_safe(c.size() * sizeof(T), split_align) * split_align);
+   {      
+      size_t data_size = sizes.first + cudf::util::round_up_safe(c.size() * sizeof(T), split_align);  
       size_t validity_size = sizes.second + (c.nullable() ? cudf::bitmask_allocation_size_bytes(c.size(), split_align) : 0);
```
You don't even need the second argument to `cudf::bitmask_allocation_size_bytes()`, since it defaults to 64.
